ARIMAX: fitting model, prediction and tests

ekote opened the pull request.

Commit:
ARIMAX: fitting model, prediction and tests

blbarker commented on src/main/scala/com/cloudera/sparkts/models/ARIMAX.scala:
```diff
@@ -0,0 +1,583 @@
+    *
+    * @param p Autoregressive order
+    * @param d Differencing order
+    * @param q Aoving average order
```
"Moving"?

ekote replied:
Done

blbarker commented on src/main/scala/com/cloudera/sparkts/models/ARIMAX.scala:
```diff
@@ -0,0 +1,583 @@
+               xregMaxLag: Int,
+               includeOriginalXreg: Boolean = true,
+               includeIntercept: Boolean = true,
+               userInitParams: Array[Double] = null): ARIMAXModel = {
```
A common pattern in scala is to use the Option[T] datatype to specify absence of value, rather than null.

blbarker commented on src/main/scala/com/cloudera/sparkts/models/ARIMAX.scala:
```diff
@@ -0,0 +1,583 @@
+      val xregVector = new BreezeDenseVector(xreg.toArray)
+      val differentialXreg = differencesOfOrderD(xregVector, d).toArray
+      val differentialXregVector = new BreezeDenseVector(differentialXreg)
+      val differentialXregMatrix = new BreezeDenseMatrix(xreg.rows, xreg.cols, differentialXregVector.toArray)
```
This code may be more readable, testable and resusable if some of these steps were broken into separate methods

blbarker commented on src/main/scala/com/cloudera/sparkts/models/ARIMAX.scala:
```diff
@@ -0,0 +1,583 @@
+    val initialGuess = new InitialGuess(initParams)
+    val maxIter = new MaxIter(10000)
+    val maxEval = new MaxEval(10000)
+    val goal = GoalType.MAXIMIZE
```
may be more readable to declare some of these inline rather than calling out separate vals.

blbarker commented on src/main/scala/com/cloudera/sparkts/models/ARIMAX.scala:
```diff
@@ -0,0 +1,583 @@
+                   val includeIntercept: Boolean = true) extends TimeSeriesModel {
+
+  /**
+    * Provided fitted values for timeseries ts as 1-step ahead forecasts, based on current
```
"Provide"

blbarker commented on src/test/resources/arimax_data.csv:
```diff
@@ -0,0 +1,258 @@
+38.3333
```
Do they already have a pattern of added csv files into test/resources?  Where did this data come from?  just need to be careful about any licensing...

Commit:
After CR

dmsuehir commented on src/main/scala/com/cloudera/sparkts/models/ARIMAX.scala:
```diff
@@ -0,0 +1,583 @@
+  *
+  * @param p Autoregressive order
+  * @param d Differencing order
+  * @param q Aoving average order
```
Moving

Commit:
CR 2

Commit:
MAX - tests and extracted XREG

Commit:
Fixed bug inside IterateARMAX

Changes:
- calculating predictors for xreg
- coefficients(0) means intercept - always - so we have to iterate from 1

ekote commented on src/main/scala/com/cloudera/sparkts/models/ARIMAX.scala:
```diff
@@ -474,41 +475,44 @@ class ARIMAXModel(
     val tsSize = ts.size
     var error = 0.0
 
+    val xregMatrix = new BreezeDenseMatrix(rows = xreg(1).length, cols = xreg.size, data = xreg.flatten)
```
xreg.rows

Commit:
Fixed wrong rows number and creating tests

blbarker commented on src/main/scala/com/cloudera/sparkts/models/ARIMAX.scala:
```diff
@@ -0,0 +1,596 @@
+      }
+    })
+    val (initialGuess, maxIter, maxEval, goal) = (new InitialGuess(initParams), new MaxIter(10000), new MaxEval(10000), GoalType.MAXIMIZE)
+    val optimal = optimizer.optimize(objFunction, gradient, goal, initialGuess, maxIter, maxEval)
```
I meant do something like:

```
val optimal = optimizer.optimize(objFunction, gradient, GoalType.MAX, new InitialGuess(initParams), new MaxIter(10000), new MaxEval(10000))
```
